python/ray/air/_internal/torch_utils.py

```diff
@@ -1,70 +1,27 @@
-import os
 import warnings
 from typing import Any, Dict, List, Optional, Union
 
 import numpy as np
 import pandas as pd
 import torch
 
-import ray
-from ray._private.accelerators.hpu import HPU_PACKAGE_AVAILABLE
+from ray.air._internal.device_manager.nvidia_gpu import CUDATorchDeviceManager
 from ray.air.util.data_batch_conversion import _unwrap_ndarray_object_type_if_needed
 
-if HPU_PACKAGE_AVAILABLE:
-    import habana_frameworks.torch.hpu as torch_hpu
-
 
 def get_devices() -> List[torch.device]:
     """Gets the correct torch device list configured for this process.
 
-    Returns a list of torch CUDA devices allocated for the current worker.
-    If no GPUs are assigned, then it returns a list with a single CPU device.
-
-    Assumes that `CUDA_VISIBLE_DEVICES` is set and is a
-    superset of the `ray.get_gpu_ids()`.
+    Returns a list of torch Accelerator devices allocated for the current worker.
+    If no Accelerators are assigned, then it returns a list with a single CPU device.
     """
-    if torch.cuda.is_available():
-        # GPU IDs are assigned by Ray after you specify "use_gpu"
-        # GPU `ray.get_gpu_ids()` may return ints or may return strings.
-        # We should always convert to strings.
-        gpu_ids = [str(id) for id in ray.get_gpu_ids()]
-
-        device_ids = []
-
-        if len(gpu_ids) > 0:
-            cuda_visible_str = os.environ.get("CUDA_VISIBLE_DEVICES", "")
-            if cuda_visible_str and cuda_visible_str != "NoDevFiles":
-                cuda_visible_list = cuda_visible_str.split(",")
-            else:
-                cuda_visible_list = []
-
-            # By default, there should only be one GPU ID if `use_gpu=True`.
-            # If there are multiple GPUs, return a list of devices.
-            # If using fractional GPUs, these IDs are not guaranteed
-            # to be unique across different processes.
-            for gpu_id in gpu_ids:
-                try:
-                    device_ids.append(cuda_visible_list.index(gpu_id))
-                except IndexError:
-                    raise RuntimeError(
-                        "CUDA_VISIBLE_DEVICES set incorrectly. "
-                        f"Got {cuda_visible_str}, expected to include {gpu_id}. "
-                        "Did you override the `CUDA_VISIBLE_DEVICES` environment"
-                        " variable? If not, please help file an issue on Github."
-                    )
+    from ray.air._internal.session import _get_session
 
-        else:
-            # If called on the driver or outside of Ray Train, return the
-            # 0th device.
-            device_ids.append(0)
-
-        devices = [torch.device(f"cuda:{device_id}") for device_id in device_ids]
-    elif HPU_PACKAGE_AVAILABLE and torch_hpu.is_available():
-        devices = [torch.device("hpu")]
+    session = _get_session()
+    if session and session.device_manager:
+        return session.device_manager.get_devices()
```

Comment:
can we put this logic in a device manager getter in `torch_device_manager.py`? so we don't have to keep checking for the existence of the session

Also, let's remove this from the `session`, since it's only initialized for the `TorchTrainer`, whereas the `session` is a shared component for all trainers:

```python
# in torch_device_manager.py

_torch_device_manager = None
_torch_device_manager_lock = threading.Lock()

def get_torch_device_manager() -> TorchDeviceManager:
    from ... import CudaTorchDeviceManager

    global _torch_device_manager

    with _torch_device_manager_lock:
        return _torch_device_manager or CudaTorchDeviceManager()


def set_torch_device_manager(device_manager):
    global _torch_device_manager
    with _torch_device_manager_lock:
        _torch_device_manager = device_manager
```

Reply:
Sure!